Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp

```diff
@@ -406,6 +413,7 @@ BytecodeGenerator::BytecodeGenerator(VM& vm, FunctionNode* functionNode, Unlinke
     //
     // Note that we intentionally enable tail call for naked constructors since it does not have special code for "return".
     , m_inTailPosition(Options::useTailCalls() && !isConstructor() && constructorKind() == ConstructorKind::None && functionNode->isStrictMode())
+    , m_inCallIgnoreResultPosition(true)
```

Comment:
Why do you default `m_inCallIgnoreResultPosition` to true here?  Why not only `SetForScope` it to true in `emitNodeInTailPositionFromExprStatementNode`?  

I'm trying to understand why you went with the model where everything ignores results by default except a few places which opts out, vs a more conservative approach where everything does not ignore results (as before) except for the place where we know that it is not needed?  What insight am I missing that tells you that it's safe and correct to go with the opt out approach?

Reply:
Because I follow to the existing `m_inTailPosition` mechanism, to avoid introducing different logic and bugs.

Reply:
After discussing this with Yusuke offline, I came to understand that:
1. Yusuke's goal is not to eliminate all unused function call results.  He's only implementing a conservative set, which happens to be a majority of naturally occurring cases.
2. The conservative set is the set of all tail calls that are not in return statements.  By definition, the only user of the result of a tail call is the enclosing return statement.  If the tail call is not in a return statement, then its result is, by definition, unused.

With that understanding, I suggest the following:

Suggestion #1: Rename `m_inCallIgnoreResultPosition` to `m_allowCallIgnoreResultOptimization`.   I think `m_inCallIgnoreResultPosition` is misleading because there are other cases of statements that may be in a position where its call result can be ignored, but yet we don't set this flag ... which is what led to my original confusion.  In contrast, `m_allowCallIgnoreResultOptimization` only says whether we allow the optimization or not, and makes no statement about whether a call is in a position where its result can be ignored or not.  I think this expresses the intent more clearly.

For that matter, I've come to see that `m_inTailPosition` is also a misnomer.  It should actually be named `m_allowTailCallOptimization` instead, but that's not related to this patch.

Suggestion #2: add a comment above this line indicating something like this:
```
// Currently, we're only conservatively allowing CallIgnoreResult optimization on tail call results that are
// not in return statements. We're not attempting to eliminate all unused call results.
```